A17-client-side-health-checking.md

```diff
@@ -0,0 +1,315 @@
+ultimately fail anyway.  And second, it is not clear what the behavior
+of `pick_first` would be for unhealthy channels; the naive approach of
+treating an unhealthy channel as having failed and disconnecting would
+both result in expensive reconnection attempts and might not actually
+cause us to select a different backend after re-resolving.
+
+The `pick_first` LB policy will unconditionally set the channel arg to
+inhibit health checking.
+
+#### `round_robin`
+
+Unhealthiness in `round_robin` will be handled in the obvious way: the
+subchannel will be considered to not be in state `READY`, and picks will
+not be assigned to it.
+
+#### `grpclb`
+
+The `grpclb` LB policy will set the channel arg to inhibit health checking
+when we are using backend addresses obtained from the balancer, on the
+assumption that the balancer is doing centralized health-checking.  The arg
+will not be set when using fallback addresses, since we want client-side
+health checking for those.
+
+### Caveats
+
+Note that health checking will use a stream on every connection.
+This stream counts toward the HTTP/2 `MAX_CONCURRENT_STREAMS` setting.
+So, for example, any gRPC client or server that sets this to 1 will not
+allow any other streams on the connection.
+
+Enabling health checks may affect the `MAX_CONNECTION_IDLE` setting.
+We do not expect this to be a problem, since we are only implementing
+client-side health checking in `round_robin`, which always attempts to
+reconnect to every backend after an idle timeout anyway.
+
+## Rationale
+
+We discussed a large number of alternative approaches.  The two most
+appealling are listed here, along with the reasons why they were not
+chosen.
+
+1. Use HTTP/2 SETTINGS type to indicate unhealthiness.  Would avoid proto
+dependency issues in core.  However, would be transport-specific and would
+not work for HTTP/2 proxies.  Also would not allow signalling different
+health status for multiple services on the same port.  Would also
+require server applications to use different APIs for the centralized
+health checking case and the client-side health checking case.
+
+2. Have server stop listening when unhealthy.  This would be challenging
+to implement in Go due to the structure of the listener API.  It would
+also not allow signalling different health status for multiple services on
+the same port.  Would also require server applications to use different
+APIs for the centralized health checking case and the client-side health
+checking case.  Would not allow individual clients to decide whether to
+do health checking.  In cases where the server's health was flapping up
+and down, would cause a lot of overhead from connections being killed
+and reestablished.
+
+## Implementation
+
+### C-core
+
+- Implement new `Watch()` method in health checking service
+  (https://github.com/grpc/grpc/pull/16351).
+- Add support for health checking in subchannel code using nanopb
+  (in progress).
+
+### Java
```

Comment:
Will also need plumbing to allow setting the tracing event.

Reply:
Added.